src/sentry/middleware/auth.py

```diff
@@ -1,23 +1,46 @@
 from __future__ import absolute_import
 
-from django.contrib.auth import middleware
+from django.contrib.auth.models import AnonymousUser
+from django.contrib.auth import get_user as auth_get_user
+from django.utils.functional import SimpleLazyObject
 
 from sentry.utils.linksign import process_signature
 from sentry.utils import auth
 
 
-class AuthenticationMiddleware(middleware.AuthenticationMiddleware):
+def get_user(request):
+    if not hasattr(request, '_cached_user'):
+        user = auth_get_user(request)
+        # If the user bound to this request matches a real user,
+        # we need to validate the session's nonce. This nonce is
+        # to make sure that the session is valid for effectively the
+        # current "version" of the user. When security related
+        # actions take place, this nonce will rotate causing a
+        # mismatch here forcing the session to be logged out and
+        # requiring re-validation.
+        if user.is_authenticated():
+            # We only need to check the nonce if there is a nonce
+            # currently set on the User. By default, the value will
+            # be None until the first action has been taken, at
+            # which point, a nonce will always be required.
+            if user.session_nonce and request.session.get('_nonce', '') != user.session_nonce:
```

Comment:
Django prefixes some internal keys with _ in similar vein to a "private variable". This is differentiated from a normal session variable that we wouldn't want someone tinkering with accidentally.